lib/datadog/di/serializer.rb

```diff
@@ -92,7 +102,14 @@ def serialize_value(value, name: nil, depth: settings.dynamic_instrumentation.ma
         when Integer, Float, TrueClass, FalseClass
           serialized.update(value: value.to_s)
         when String, Symbol
-          value = value.to_s
+          value = case value
+          when String
+            # This is the only place where we duplicate the value, currently.
+            # All other values are immutable primitives (e.g. numbers).
+            value.dup
+          else
+            value.to_s
+          end
           max = settings.dynamic_instrumentation.max_capture_string_length
           if value.length > max
             serialized.update(truncated: true, size: value.length)
```

Comment:
May be worth using https://github.com/DataDog/dd-trace-rb/blob/master/lib/datadog/core/utils/safe_dup.rb or `-v` directly? :thinking: 

Also, it may be worth checking if truncation needs to happen first, otherwise we'll be creating yet another copy for the truncation :eyes:

Reply:
Good point, if the string is to be truncated there is no need to dup it, and for frozen strings there is also no need to dup since they can't be mutated.

Reply:
Done.